core/src/qos_service.rs

```diff
@@ -144,34 +144,148 @@ impl QosService {
     ) {
         while running_flag.load(Ordering::Relaxed) {
-            metrics.report(reporting_duration_ms);
+            for bank in report_receiver.try_iter() {
+                metrics.report(bank.slot());
+            }
             thread::sleep(Duration::from_millis(100));
         }
     }
 }
 
 #[derive(Default)]
 struct QosServiceMetrics {
-    last_report: AtomicInterval,
+    // bankign_stage creates one qos_service instance per working threads, which is uniquely
+    // identified by id. This field allows to categorize metrics for gossip votes, TPU votes
+    // and other transactions.
+    id: u32,
+
+    // aggregate metrics per slot
+    slot: AtomicU64,
+
+    // accumulated number of live packets TPU received from verified receiver for processing.
+    tpu_ingested_packets_count: AtomicU64,
+
+    // accumulated number of live packets TPU put into buffer due to no active bank.
+    tpu_buffered_packets_count: AtomicU64,
+
+    // accumulated number of verified txs, which excludes unsanitized transactions and
+    // non-vote transactions when in vote-only mode from ingested packets
+    verified_txs_count: AtomicU64,
+
+    // accumulated number of transactions been processed, includes those landed and those to be
+    // retriued (due to AccountInUse, and other QoS related reasons)
+    processed_txs_count: AtomicU64,
+
+    // accumulated number of transactions buffered for retry, often due to AccountInUse and QoS
+    // reasons, includes retried_txs_per_block_limit_count and retried_txs_per_account_limit_count
+    retryable_txs_count: AtomicU64,
+
+    // accumulated time in micro-sec spent in computing transaction cost. It is the main performance
+    // overhead introduced by cost_model
     compute_cost_time: AtomicU64,
+
+    // total nummber of transactions in the reporting period to be computed for theit cost. It is
+    // usually the number of sanitized transactions leader receives.
     compute_cost_count: AtomicU64,
+
+    // acumulated time in micro-sec spent in tracking each bank's cost. It is the second part of
+    // overhead introduced
     cost_tracking_time: AtomicU64,
+
+    // number of transactions to be included in blocks
     selected_txs_count: AtomicU64,
+
+    // number of transactions to be queued for retry due to its potential to breach block limit
     retried_txs_per_block_limit_count: AtomicU64,
+
+    // number of transactions to be queued for retry due to its potential to breach writable
+    // account limit
     retried_txs_per_account_limit_count: AtomicU64,
 }
 
 impl QosServiceMetrics {
-    pub fn report(&self, report_interval_ms: u64) {
-        if self.last_report.should_update(report_interval_ms) {
+    pub fn new(id: u32) -> Self {
+        QosServiceMetrics {
+            id,
+            ..QosServiceMetrics::default()
+        }
+    }
+
+    pub fn report(&self, bank_slot: Slot) {
+        if bank_slot != self.slot.load(Ordering::Relaxed) {
```

Comment:
The flow as I understand it is

1. Every banking thread sends a `Bank` over a channel at the end of `process_and_record_transactions()` via a call to `report_metrics()` https://github.com/solana-labs/solana/blob/65a145db2f7d2e69c6ce5df4c8da02adc44c4993/core/src/banking_stage.rs#L968
2. There exists a unique `QosService` thread for that particular banking thread that will process that bank from the channel and call this function `report`

Might be good to add a comment for this at the top of this service to summarize this

Reply:
Yes, will do